webapp/src/ui/courses.js

```diff
@@ -1,100 +1,93 @@
-} from '@material-ui/core'
-import { connect } from 'react-redux'
-import { getCourses } from '../actions/courses'
-
-const Span = styled.span`
-  font-size: 10px;
-  text-align: left;
-  display: flex;
-  justify-content: left,
-  align-content: left,
-  text-align: left
-`;
-
-
-class Courses extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      filtered: [],
-      loading: true,
-      query: ""
-    }
-  }
-
-  componentDidMount() {
-    this.props.getCourses()
-  }
-
-  filterCourses = (filterKey) => {
-    if(!this.props.courses.results) return [];
-    let filteredCourses = this.props.courses.results;
-    let searchFilter = filterKey.split(' ');
-    filteredCourses = filteredCourses.filter(course => {
-      let entry = Object.values(course);
-      let entryString = entry.slice(0, entry.length)
-      let modifiedEntryString = entryString.join().toLowerCase();
-      return searchFilter.every(key => {
-        return modifiedEntryString.includes(key.toLowerCase())
-      });
-    });
-    return filteredCourses;
-  }
-
-  handleChange = (event: object) => {
-    let searchKey = event.target.value.toLowerCase();
-    this.setState({ query: searchKey });
-    let data = this.filterCourses(searchKey);
-    this.setState({ filtered: data })
-  }
-
-
-  render() {
-    const courses = this.props.courses;
-    const state = this.state;
-    const filtered = state.filtered;
-    const totalData = courses.total;
-    let data = courses.results;
-    if(filtered.length > 0 && filtered.length < data.length) {
-      data = filtered;
-    }
-    return (
-      <div className="courses">
-        <Span>
-          Matched {filtered.length} results
-        </Span>
-        <Span> Got {totalData ? totalData : 0} results </Span>
-        <Input
-          value={state.query}
-          onChange={(e) => this.handleChange(e)}
-          fullWidth={true}
-          placeholder="Search..."
-          autoFocus={true}
-        />
-        {!this.props.courses.loading
-          ? <CoursesList data={data} />
-          : <LinearProgress />
-        }
-      </div>
-    );
-  }
-}
-
-function mapStateToProps(state) {
-  return state
-}
-
-const mapDispatchToProps = {
-  getCourses
-}
-
-export default connect(
-  mapStateToProps,
-  mapDispatchToProps
-)(Courses)
+import React, {useState, useEffect} from 'react';
+import {makeStyles} from '@material-ui/core/styles';
+import {InputBase, Backdrop, CircularProgress, Paper, Box, InputAdornment} from '@material-ui/core';
+import SearchIcon from '@material-ui/icons/Search';
+import {useDebounce} from 'use-debounce';
+import CoursesList from './components/courses-list';
+import {useDispatch, useSelector} from 'react-redux';
+import {getCourses} from '../store/actions/courses';
+
+const useStyles = makeStyles(() => ({
+	backdrop: {
+		zIndex: 1000
+	}
+}));
+
+const Courses = () => {
+	const dispatch = useDispatch();
+
+	// Update courses on load
+	useEffect(() => {
```

Comment:
If we want to use this pattern, we should consider re-implementing thunk and start using selectors as well
See:
https://github.com/reduxjs/redux/issues/1171
https://github.com/reduxjs/redux-thunk#why-do-i-need-this

Also consider maybe looking into React's Lazy loading
https://reactjs.org/docs/code-splitting.html#reactlazy

It sort of does what we're doing already with the "if no data display a loading screen" logic but makes it require less boiler plate.

Reply:
I read through a couple different posts about redux thunk patterns ([this was really helpful](https://daveceddia.com/what-is-a-thunk/)), and I now have a much better understand of it than when I removed it. 😛 As far as I can tell, it doesn't make sense to use `redux-thunk` *until* we want to write our own thunks, as asynchronous network requests are currently handled by `redux-api-middleware`.

Where would you want to use selectors that we're not already? The only place we pull data from the store for display currently is here:
```js
const courses = useSelector(state => state.courses);
```

I'm not sure if lazy loading does what you think, AFAIK it's made more for code splitting (for example, don't load JS for page X until user hovers over link to page X).

Reply:
More what I was talking about with the selectors is breaking them apart from the main components, and placing them in the directory tree elsewhere. More on that point was there any specific reason the directory tree is now:
```
.
├── config.js
├── index.js
├── store
│   ├── actions
│   │   └── courses.js
│   ├── configure-store.js
│   └── reducers
│       ├── courses.js
│       └── index.js
└── ui
    ├── components
    │   └── courses-list.js
    ├── courses.js
    └── page.js
```
?

React Lazy was just something cool I was looking at and thinking about switching to in other projects rather than custom loading logic.

You are right in that we don't actually need to use redux-thunk until we want to write our own custom thunks. Thank you for the resource.

Reply:
Got it.

No, I just used the existing structure, and tried to organize Redux a bit by putting it all under `store/`. I'm open to other ways of structuring if you have ideas.  I usually do something like
- pages
	- _app.js (container)
 	- index.js (home page)
	- etc
- components
- lib

Reply:
In the past, I've used:

- src/
  - _some app.js
    - components/
    - App.js (Name doesn't really matter)
    - reducers.js
    - selectors.js
    - actions.js
  - _some app2.js
  - components/
etc.

Though for a small project it might not be necessary to break it up as much. It doesn't really make too much sense to stuff everything redux under `store` in my opinion.

Reply:
Is this more what you were thinking? 
db591d7